### PR TITLE
Add:gui_internal:Support native on-screen keyboard (currently Android…

### PR DESCRIPTION
**Pull request by mvglasow**
… only)

Squashed commit of the following:

commit 2535b0f7ebc84a2670ad5bba31ffb444a5082311
Author: mvglasow <michael -at- vonglasow.com>
Date:   Thu Nov 19 09:24:51 2015 +0100

    Refactor:port_android:Remove forgotten unneeded import

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

commit 680b83322abde1ff72b1189d969e5fa62490c7c0
Merge: 0b1d8e4 83b07f8
Author: mvglasow <michael -at- vonglasow.com>
Date:   Thu Nov 19 09:05:19 2015 +0100

    Merge branch 'trunk' into trac712

commit 0b1d8e44aead719d58513096e0754730951abb7b
Author: mvglasow <michael -at- vonglasow.com>
Date:   Wed Nov 18 14:30:09 2015 +0100

    Refactor:gui_internal:Code cleanup and documentation

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

commit 0b699dbe577d8a49f101513e2a6dbe12cc454dd3
Author: mvglasow <michael -at- vonglasow.com>
Date:   Tue Nov 17 16:40:50 2015 +0100

    Refactor:port_android:some comments

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

commit ab5bdefbf51a0eaf9ce9ee697a0ce38467a86003
Author: mvglasow <michael -at- vonglasow.com>
Date:   Tue Nov 17 16:36:50 2015 +0100

    Add:gui_internal:Support mode for native virtual keyboard

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

commit 86dec4dd287d29b5666cefa24b4d5f5aa67671cb
Author: mvglasow <michael -at- vonglasow.com>
Date:   Mon Nov 16 16:36:21 2015 +0100

    Fix:port_android:Restore soft keyboard when resuming activity

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

commit d3c698fb53cf3ab5f93e50c7f0287db4c025a6a2
Author: mvglasow <michael -at- vonglasow.com>
Date:   Sun Nov 15 23:05:24 2015 +0100

    Fix:gui_internal:Further native keyboard refinements

    Allow graphics plugin to specify size occupied by keyboard
    Create placeholder for Android keyboard
    Show Android keyboard in landscape mode
    Skip keyboard logic if hardware keyboard is present on Android

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

commit a857267232a243f58ffdd01f5fe352d7d3e4885a
Author: mvglasow <michael -at- vonglasow.com>
Date:   Fri Nov 13 23:49:19 2015 +0100

    Add:gui_internal:Rudimentary native on-screen keyboard implementation

    Signed-off-by: mvglasow <michael -at- vonglasow.com>

### CODE REVIEW

**Comment by sleske** [comment]
Is it really necessary to sqash all the commits into one? Why not merge the commits individually, or rebase them on top of trunk?
I find squashed commits hard to work with, hard to read, they make bisecting impossible, etc.
If there is a technical reason, maybe we can find a solution?

**Comment by mvglasow** [comment]
TL;DR: If you'd prefer to see the individual commits, look at https://github.com/navit-gps/navit/compare/trunk...mvglasow:trac712 – I could merge that as well.

Actually the reason why is squashed those commits has more to do with legibility than with technical constraints. This goes especially for code history after the merge – the merge request still quite legible because it lists only the commits which are not yet in trunk, but the merged history will additionally list all merges between branches.

I rely quite heavily on commits for my own development. A typical scenario may look like this:
* I start by digging my way through the code and find that many of the functions I might need for my new feature are not documented, so I add a bunch of Doxygen comments. When I'm about to write the first line of actual code, I first commit the documentation back to trunk.
* Now I create a new feature branch `feature42` and start coding. As I go along, I notice more functions that need documentation (that is, I might need to rely on them but they are not documented and I do not understand them right away, so I analyze them) and add what is missing. At this point new code and documentation get hard to separate – I don't want to commit the code because it may be so incomplete that Navit may not even build, but I also need the documentation before I continue coding.
* After writing some more code, I discover a bug in a different place which I need to fix urgently. The bug has nothing to do with my feature, so I need to switch to `trunk` or a new feature branch. So I either stash my current working state – or, because multiple stashes get messy to work with, I just commit them with the message "stash", knowing that I'm going to amend this commit later – and fix the bug, ultimately committing it back to trunk.
* Eventually I switch back to my feature branch and continue coding up to a point at which I have a rudimentary new feature, so I commit that (amending the earlier "stash" commit – not an issue as long as that commit has not yet made it anywhere official). I put the code up for discussion and collect some feedback on it.
* In the meantime, someone comes back to me with feedback on another branch, `feature22` that I've put up for discussion a while back, and suggests some changes. I realize that these changes are going to rely on some functions for which I have worked out the documentation in `feature42` but it's in a commit mixed with code changes (see three points before). So I cherry-pick that commit into my local copy of `trunk`, amend it by undoing the code changes and leaving only the documentation, then push it back into the upstream repo. A merge of `trunk` into `feature22` and I can finally finish my work there.
* Back to `feature42`: to be safe, I merge `trunk` back into `feature42` – else merge hell might ensue if I make any further edits to the documentation I cherry-picked into `trunk`.
* Now I get some feedback on `feature42` and realize I need to throw away and rewrite a good portion of the code I wrote for that feature.
Eventually I finish work on `feature42`, but the commit history is quite a mess, showing documentation that I have already pushed to `trunk` in the meantime, as well as a bunch of merge operations, and commits that (at least partially) cancel each other out. The "stash-and-amend" tactic will not work if for some reason I need to merge before I can fully amend the "stash" commit. As a result, some of the intermediate commits may not even have produced workable code, defeating bisect.

A real-world examples that has suffered from the above scenario is https://github.com/mvglasow/navit/tree/android-laf-trac1325. Feel free to clone it and view it in `git gui`. Github makes things look a bit tidier by arranging commits chronologically, but this just hides the mess underneath.

Therefore, as long as the changes are all related to a single feature and not overly extensive, I personally find a squashed commit easier to read than the whole history. In fact, this is an approach I've learned from the guys over at Mapsforge, who rejected a merge request of some 10 commits that I'd filed, asking me to squash them into a single commit. We've done something similar in Navit in the past – together with jandegr and robotaxi I spent some six months working on HighFive, and when it was time to merge, I asked kazer how he'd prefer it and he suggested that I do it in a single commit. We still had SVN back then – now, with the potential of git at my fingertips, I opted for the compromise of squashing everything into one commit but keeping references to the individual commits in the commit log.

I'm open to discussion, though – and if you have a good suggestion how I can avoid the above scenario from turning the commit history into a complete mess, I'd love to hear it.

**Comment by sleske** [comment]
Thanks for the detalied explanations - I did not expect so much :+1: . I'll try to go through it all and follow up soon. In the meantime:
Would you mind if I copy (or would you copy) your post to the forum, to continue discussion there? This is really about coding / development style, not specific to this pull request.

**Comment by mvglasow** [comment]
I've created a discussion here: https://forum.navit-project.org/viewtopic.php?f=19&t=590

**Comment by sleske** [comment]
Thanks! I'll reply over there.

**Comment by mvglasow** [comment]
I've rebased my original branch and squashed what needed to be squashed, reopened as #50, closing this one
